source/client-side-encryption/client-side-encryption.rst

```diff
@@ -286,13 +289,86 @@ Data keys are stored as documents in a special MongoDB collection. Data
 keys are protected with encryption by a KMS provider (AWS KMS, Azure key
 vault, GCP KMS, or a local master key).
 
+metadataClient
+^^^^^^^^^^^^^^
+A client used to run ``listCollections`` to determine whether a collection has
+an associated JSON schema with encrypted fields specified.
+
+If a ``metadataClient`` is not passed in, and
+``bypassAutomaticEncryption=false``, the ``metadataClient`` is set to an
+internal ``MongoClient``. See `keyVaultClient, metadataClient, and the internal
+MongoClient`_ for configuration behavior.
+
+Drivers MUST document this behavior, using the following as a template:
+
+   If ``metadataClient`` is not passed as an option, and
+   ``bypassAutomaticEncryption=false``, a separate ``MongoClient`` is used
+   internally. It is configured using the options of the parent ``MongoClient``
+   with the ``AutoEncryptionOpts`` omitted.
+
+See `What's the deal with metadataClient, keyVaultClient, and the internal client?`_
+
 keyVaultClient
 ^^^^^^^^^^^^^^
 The key vault collection is assumed to reside on the same MongoDB
 cluster as indicated by the connecting URI. But the optional
 keyVaultClient can be used to route data key queries to a separate
 MongoDB cluster.
 
+If a ``keyVaultClient`` is not passed, the ``keyVaultClient`` is set to an
+internal ``MongoClient``. See `keyVaultClient, metadataClient, and the internal
+MongoClient`_ for configuration behavior.
+
+Drivers MUST document this behavior, using the following as a template:
+
+   If ``keyVaultClient`` is not passed as an option, a separate ``MongoClient``
+   is used internally. It is configured using the options of the parent
+   ``MongoClient`` with the ``AutoEncryptionOpts`` omitted.
+
+See `What's the deal with metadataClient, keyVaultClient, and the internal client?`_
+
+keyVaultClient, metadataClient, and the internal MongoClient
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+If a ``keyVaultClient`` is not passed, it is set to an internal ``MongoClient``.
+
+If a ``metadataClient`` is not passed, and ``bypassAutomaticEncryption=false``,
+it is set to an internal ``MongoClient``. If ``bypassAutomaticEncryption=true``,
+the option is ignored since ``listCollections`` is only run during automatic
+encryption.
+
+The ``keyVaultClient`` and ``metadataClient`` MUST be set to the same internal
+``MongoClient`` if neither are passed in and ``bypassAutomaticEncryption=false``.
+
+The internal ``MongoClient`` MUST be configured with the same options as the
+parent ``MongoClient`` with the ``AutoEncryptionOpts`` excluded.
+
+The following pseudo-code describes the configuration behavior:
+
+.. code::
+
+   def getOrCreateInternalClient (client, clientOpts):
```

Comment:
Food for thought: from my understanding, the problem arises when specifying `maxPoolSize` in the original client options. If these are not passed, is there a reason for creating a separate client? Could we reuse the original client in those cases to save resources?

Reply:
Yes, we could reuse the original client.

It might not be a less important use case to account for. `maxPoolSize` defaults to `100`. If a user explicitly opts into `maxPoolSize=0`, they may not be concerned with extra connections. OTOH, they may be configuring with a large `minPoolSize` as well.

I'll account for this if it does not add much more complexity.

Reply:
This has been accounted for in 68374f8